### PR TITLE
[TRITON] Add script to select Triton tests based on diff content

### PR DESCRIPTION
**Pull request by brunomazzottiamd**
## Motivation

Triton test step of AITER CI is taking too long and becoming a bottleneck for PR merging. We should programmatically select which Triton tests to run, based on the diff content of a given PR. The full test suite can be executed periodically on `main` branch only.

This PR proposes a Python script that given

* a source branch with changes to be merged
* a target branch to merge into
* the AITER codebase

automatically selects which Triton tests validate the changes to be merged.

## Technical Details

Script help text:

```
$ python .github/scripts/select_triton_tests.py --help
usage: select_triton_tests.py [-h] [-s SOURCE] [-t TARGET] [-v ENV_VAR] [-f ENV_FILE] [-l {critical,error,warning,info,debug,off}]

select which Triton tests to run based on git diff

options:
  -h, --help            show this help message and exit
  -s SOURCE, --source SOURCE
                        source branch, defaults to current branch
  -t TARGET, --target TARGET
                        target branch, defaults to main
  -v ENV_VAR, --env-var ENV_VAR
                        environment variable to store which tests to run, defaults to TRITON_TEST
  -f ENV_FILE, --env-file ENV_FILE
                        environment file to write, won't write anything if absent
  -l {critical,error,warning,info,debug,off}, --log-level {critical,error,warning,info,debug,off}
                        log level to enable (default: info)
```

Implementation details:

* Uses Python [`subprocess`](https://docs.python.org/3/library/subprocess.html) module to run `git` commands and find out what has changed between source and target branches.
* Uses Python [`pathlib`](https://docs.python.org/3/library/pathlib.html) module to list all Triton source files in AITER code base (kernels, kernel configurations, unit tests, benchmark scripts).
* Uses Python [`ast`](https://docs.python.org/3/library/ast.html) module to recursively parse the Triton source files, tracking all dependency relations. These dependency relations can be between two source files or between a kernel configuration file and a kernel source file.
* The dependency relations are encoded in a [`networkx`](https://networkx.org/en/) directed graph. Later, the graph is traversed starting from the diff content until we reach unit tests. All unit tests that are reachable should be executed to validate the proposed changes.
* No code is executed, the entire construction of the dependency graph is done through static analysis.

## Test Plan

Tested locally with these scenarios:

* Test case A - changed the following Triton source files:
  * `aiter/ops/triton/_triton_kernels/mha.py` → MHA forward kernel
  * `aiter/ops/triton/_triton_kernels/mha_onekernel_bwd.py` → MHA backward kernel
  * `aiter/ops/triton/_triton_kernels/rope.py` → RoPE kernel
  * `aiter/ops/triton/configs/gfx942-GMM.json` → GMM kernel configuration
  * `op_tests/op_benchmarks/triton/bench_topk.py` → Top-k benchmark script
* Test case B - changed files other than Triton sources.
* Test case C - this feature branch, i.e. `bmazzott/run-triton-tests-selectively`, which has changes in:
  * `aiter/ops/triton/_triton_kernels/fused_gemm_afp4wfp4_a16w16.py` → standardization of `get_gemm_config` arguments
  * `aiter/ops/triton/_triton_kernels/fused_gemm_afp4wfp4_mul_add.py` → standardization of `get_gemm_config` arguments 
  * `aiter/ops/triton/_triton_kernels/gemm_afp4wfp4.py` → standardization of `get_gemm_config` arguments
  * `aiter/ops/triton/_triton_kernels/gmm.py` → standardization of kernel config file pattern
* Test case D - changed an untested Triton kernel:
  * `aiter/ops/triton/_triton_kernels/pod_attention.py` → untested kernel

## Test Result

Summarized script output:

Test case A:

```
INFO|There are 5 Triton source files in the diff:
INFO|* aiter\ops\triton\_triton_kernels\mha.py
INFO|* aiter\ops\triton\_triton_kernels\mha_onekernel_bwd.py
INFO|* aiter\ops\triton\_triton_kernels\rope.py
INFO|* aiter\ops\triton\configs\gfx942-GMM.json
INFO|* op_tests\op_benchmarks\triton\bench_topk.py
INFO|There are 7 tests reachable from the Triton diff:
INFO|* op_tests\triton_tests\attention\test_mha.py
INFO|* op_tests\triton_tests\fusions\test_fused_kv_cache.py
INFO|* op_tests\triton_tests\fusions\test_fused_qk_concat.py
INFO|* op_tests\triton_tests\rope\test_fused_qkv_split_qk_rope.py
INFO|* op_tests\triton_tests\rope\test_rope.py
INFO|* op_tests\triton_tests\test_gmm.py
INFO|* op_tests\triton_tests\test_topk.py
INFO|Finished, execution took 7.41 seconds.
```

Test case B:

```
INFO|There are no Triton source files in diff, there's no need to run Triton tests.
```

Test case C:

```
INFO|There are 4 Triton source files in the diff:
INFO|* aiter\ops\triton\_triton_kernels\fused_gemm_afp4wfp4_a16w16.py
INFO|* aiter\ops\triton\_triton_kernels\fused_gemm_afp4wfp4_mul_add.py
INFO|* aiter\ops\triton\_triton_kernels\gemm_afp4wfp4.py
INFO|* aiter\ops\triton\_triton_kernels\gmm.py
INFO|There are 9 tests reachable from the Triton diff:
INFO|* op_tests\triton_tests\gemm\basic\test_gemm_a16wfp4.py
INFO|* op_tests\triton_tests\gemm\basic\test_gemm_afp4wfp4.py
INFO|* op_tests\triton_tests\gemm\batched\test_batched_gemm_a16wfp4.py
INFO|* op_tests\triton_tests\gemm\fused\test_fused_gemm_afp4wfp4_a16w16.py
INFO|* op_tests\triton_tests\gemm\fused\test_fused_gemm_afp4wfp4_mul_add.py
INFO|* op_tests\triton_tests\gemm\fused\test_fused_gemm_afp4wfp4_split_cat.py
INFO|* op_tests\triton_tests\quant\test_fused_mxfp4_quant.py
INFO|* op_tests\triton_tests\test_activation.py
INFO|* op_tests\triton_tests\test_gmm.py
INFO|Finished, execution took 2.36 seconds.
```

Test case D:

```
INFO|There is 1 Triton source file in the diff:
INFO|* aiter\ops\triton\_triton_kernels\pod_attention.py
WARNING|Triton source file [aiter\ops\triton\_triton_kernels\pod_attention.py] isn't in the dependency graph, it's unreachable.
WARNING|Couldn't find any test file related to Triton diff.
WARNING|Please check test selection script, there might be a bug in it.
WARNING|Please check Triton code base, there may be untested kernels.
INFO|Finished, execution took 2.34 seconds.
```
## Triton test execution speedup

Using this PR as an example:
* Source branch: `bmazzott/run-triton-tests-selectively`
* Target branch: `main`

| Metric       | Full Triton test suite | Selected Triton tests based on PR diff content | Comment                |
|:-------------|-----------------------:|-----------------------------------------------:|:-----------------------|
| Test files   |                     60 |                                              9 | 85% fewer test files   |
| Test cases   |                 159062 |                                          17096 | 89.3% fewer test cases |
| Running time |      2h 2m 19s (7322s) |                                  5m 22s (322s) | 22.7 speedup           |

## TODO before merging

* [x] Address all comments made by Bruno Mazzotti, the PR author.
* [x] Find a way to integrate the script with CI infrastructure and GitHub Actions. I think AITER CI team (Xin Huang + Leonid Drozdov) can help with this.

## IMPORTANT NOTICE!

* The sole goal of this PR is to **just run** the script on CI, on every subsequent PR to be merged.
* Under no circumstances, should any script execution error be considered a CI failure.
* For now, we **should not** trust the script output, just monitor it and fix possible issues over time. Only when we are confident in the result, should it be used to select which tests to run.
* Another **must have** condition to use the script to drive test selection is to have a periodic run of the full Triton test suite on `main` branch.

## Submission Checklist

- [X] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: aiter/jit/core.py [hunk 1]
```diff
@@ -18,12 +18,10 @@
 
 from packaging.version import Version, parse
 
-this_dir = os.path.dirname(os.path.abspath(__file__))
-sys.path.insert(0, f"{this_dir}/utils/")
-from chip_info import get_gfx, get_gfx_list
-from cpp_extension import _jit_compile, get_hip_version
-from file_baton import FileBaton
-from torch_guard import torch_compile_guard  # noqa: E402
+from .utils.chip_info import get_gfx, get_gfx_list
```
Comment (brunomazzottiamd) [inline aiter/jit/core.py]: **FIXME:** Changes in `aiter/jit/core.py` broke AITER compilation,

File: .github/scripts/select_triton_tests.py [hunk 2]
```diff
@@ -0,0 +1,740 @@
...
+import logging
+import os
+import shlex
+import subprocess
+import sys
+import time
+from pathlib import Path
+
+# Third party libraries.
+import networkx as nx
```
Comment (brunomazzottiamd) [inline .github/scripts/select_triton_tests.py:20]: `networkx` is a third-party dependency. How can we be sure to have it installed before running the script?

File: .github/scripts/select_triton_tests.py [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+
+
+def git_check_branch(branch: str) -> None:
+    if git(f"rev-parse --verify --quiet {branch}", check=False).returncode != 0:
+        logging.critical("Branch [%s] doesn't exist.", branch)
+        sys.exit(1)
+
+
+def git_filename_diff(source_branch: str, target_branch: str) -> set[Path]:
+    # FIXME: The statement bellow only works if the script is invoked in AITER
```
Comment (brunomazzottiamd) [inline .github/scripts/select_triton_tests.py]: Resolve this **FIXME** comment before merging.

File: .github/scripts/select_triton_tests.py (hunk 2 above)
Comment (gyohuangxin) [reply .github/scripts/select_triton_tests.py:20]: You can add it here: https://github.com/ROCm/aiter/blob/main/.github/scripts/build_aiter_triton.sh#L28 to install it before running the Triton tests in CI.

File: aiter/jit/core.py (hunk 1 above)
Comment (brunomazzottiamd) [reply aiter/jit/core.py]: Reverting these changes fixed AITER compilation.

File: .github/scripts/select_triton_tests.py (hunk 2 above)
Comment (brunomazzottiamd) [reply .github/scripts/select_triton_tests.py:20]: It's done, let's see if CI runs fine with this change.

File: .github/scripts/select_triton_tests.py (hunk 3 above)
Comment (brunomazzottiamd) [reply .github/scripts/select_triton_tests.py]: Fixed.

File: .github/scripts/select_triton_tests.py [hunk 4]
```diff
@@ -0,0 +1,806 @@
...
+
+def get_filename_diff(source_branch: str | None, target_branch: str) -> set[Path]:
+    if source_branch is None:
+        source_branch = git_current_branch()
+        logging.info(
+            "Source branch wasn't provided, using current branch [%s] as source branch.",
+            source_branch,
+        )
+    else:
+        git_check_branch(source_branch)
```
Comment (gyohuangxin) [inline .github/scripts/select_triton_tests.py:288]: I'm not sure the reason CI didn't find the branch, could you add some debug steps like `git branch -r` and `git branch`?

Comment (brunomazzottiamd) [reply .github/scripts/select_triton_tests.py:288]: Fine, I'll add some `git` commands before running the script to see what's going on. Thank you for the suggestion.

Comment (brunomazzottiamd) [reply .github/scripts/select_triton_tests.py:288]: I couldn't run any `git` command in the container with `docker exec -w /workspace triton_test git ...`.

I was getting the following error:
```
fatal: detected dubious ownership in repository at '/workspace'
To add an exception for this directory, call:
	git config --global --add safe.directory /workspace
```

This error happens due to a security feature that occurs when the repository's owner is different from the user running the `git` command. This prevents unauthorized scripts from running in repositories not owned by the current user. You can resolve this error by either explicitly marking the directory as safe in your Git configuration or by changing the directory's ownership to your current user.

After running `git config --global --add safe.directory /workspace` the test selection script is working. However, I'm not sure if this is the best solution! I'm open to suggestions!

Comment (brunomazzottiamd) [reply .github/scripts/select_triton_tests.py:288]: For the sake of reference: https://github.com/ROCm/aiter/blob/c43f997c60c4b750de328f0bbe23730f0dbbfd13/.github/workflows/triton-test.yaml#L126-L127

File: .github/workflows/triton-test.yaml [hunk 5]
```diff
@@ -39,8 +41,23 @@ jobs:
...
+          ref: ${{ github.head_ref }}
+          fetch-depth: 1
           submodules: 'recursive'
 
+      - name: Fetch target branch
+        run: |
+          git fetch origin --no-tags \
+            +refs/heads/${{ github.base_ref }}:refs/remotes/origin/${{ github.base_ref }}
+
+      # TODO: Evaluate if this step should be removed before merging.
```
Comment (brunomazzottiamd) [inline .github/workflows/triton-test.yaml]: Resolve before merging.

File: .github/workflows/triton-test.yaml [hunk 6]
```diff
@@ -98,12 +115,30 @@ jobs:
           triton_test \
           pip install pytest
 
+      # TODO: Evaluate if the security exception is the best way to fix the script execution.
```
Comment (brunomazzottiamd) [inline .github/workflows/triton-test.yaml]: Resolve before merging.

File: .github/workflows/triton-test.yaml [hunk 7]
```diff
@@ -98,12 +115,30 @@ jobs:
...
+          set -ex
+          ENV_FILE=$(docker exec -w /workspace triton_test \
+            mktemp /workspace/github_env.XXXXXXXXXX.tmp)
+          docker exec -w /workspace triton_test \
+            git config --global --add safe.directory /workspace
+          docker exec -w /workspace triton_test \
+            python .github/scripts/select_triton_tests.py \
+              --source '${{ github.head_ref }}' --target 'remotes/origin/${{ github.base_ref }}' \
+              --env-var TRITON_TEST --env-file "${ENV_FILE}"
+          # TODO: Comment the follwing command before merging.
```
Comment (brunomazzottiamd) [inline .github/workflows/triton-test.yaml]: Resolve before merging, i.e. comment the following line.

**Comment by brunomazzottiamd** [comment]
Marked old discussions as resolved, hiding them.

File: .github/workflows/triton-test.yaml [hunk 8]
```diff
@@ -98,12 +115,30 @@ jobs:
           triton_test \
           pip install pytest
 
+      # TODO: Evaluate if the security exception is the best way to fix the script execution.
+      # TODO: Uncomment [docker exec -w /workspace triton_test cat "${ENV_FILE}" >> "${GITHUB_ENV}"]
+      #       command to enable test selection.
+      - name: Triton Test Selection Script
```
Comment (gyohuangxin) [inline .github/workflows/triton-test.yaml:106]: As we will always run all tests in main branch, we should skip this step when it's a main branch. 

```suggestion
      - name: Triton Test Selection Script
        if: ${{ github.ref != 'refs/heads/main' }}
```

File: .github/workflows/triton-test.yaml [hunk 9]
```diff
@@ -98,12 +115,30 @@ jobs:
...
+          docker exec -w /workspace triton_test cat "${ENV_FILE}" >> "${GITHUB_ENV}"
+          docker exec -w /workspace triton_test rm "${ENV_FILE}"
+
       - name: Triton Tests
         run: |
           set -ex
           echo "Running Triton Tests..."
           docker exec -w /workspace triton_test mkdir -p test-reports
-          docker exec -w /workspace triton_test pytest -v ${{ env.TRITON_TEST }} --junitxml=test-reports/triton.xml
+          docker exec -w /workspace triton_test pytest -v ${TRITON_TEST} --junitxml=test-reports/triton.xml
```
Comment (gyohuangxin) [inline .github/workflows/triton-test.yaml:131]: As we skip the tests on the main branch, we need to add another step.
`echo ${{ env.TRITON_TEST }} >> "${GITHUB_ENV}"`

File: .github/workflows/triton-test.yaml (hunk 8 above)
Comment (gyohuangxin) [reply .github/workflows/triton-test.yaml:106]: Regarding forked PRs, I think the simplest approach is to still run the full test suite. So we can change it to the following:
```suggestion
      - name: Triton Test Selection Script
        if: ${{ github.ref != 'refs/heads/main' && !github.event.pull_request.head.repo.fork }}
```

**Comment by brunomazzottiamd** [comment]
Forked PR test being done in https://github.com/ROCm/aiter/pull/1804.

File: .github/workflows/triton-test.yaml (hunk 8 above)
Comment (brunomazzottiamd) [reply .github/workflows/triton-test.yaml:106]: I'm trying to enable forked PRs, but I'm having a hard time... I'll try a bit more today, if I can't get it to work then I'll add the `if` statement as you suggested.

File: .github/workflows/triton-test.yaml (hunk 9 above)
Comment (brunomazzottiamd) [reply .github/workflows/triton-test.yaml:131]: From my previous tests, I think this isn't required. `${TRITON_TEST}` initial value is `${{ env.TRITON_TEST }}`, if we don't mess with `${GITHUB_ENV}` file it should be fine. I'll perform more tests later today just to confirm my hypothesis.

File: .github/workflows/triton-test.yaml (hunk 8 above)
Comment (brunomazzottiamd) [reply .github/workflows/triton-test.yaml:106]: I did this:
```yaml
      - name: Run a step only if it's from a fork
        if: ${{ github.event.pull_request.head.repo.fork }}
        run: echo "This PR is from a fork."

      - name: Run a step only if it's from the same repo (not a fork)
        if: ${{ !github.event.pull_request.head.repo.fork }}
        run: echo "This PR is from an internal branch."
```
It's always printing `This PR is from a fork.`, for https://github.com/ROCm/aiter/pull/1682 (not a fork) and https://github.com/ROCm/aiter/pull/1804 (a fork). What's going on? I'm very confused...

Comment (brunomazzottiamd) [reply .github/workflows/triton-test.yaml:106]: * Working for forked PRs.
* Implemented conditional skip for `main` branch.

File: .github/workflows/triton-test.yaml (hunk 9 above)
Comment (brunomazzottiamd) [reply .github/workflows/triton-test.yaml:131]: It's confirmed: `${TRITON_TEST}` initial value is `${{ env.TRITON_TEST }}`. We don't need to write to `${GITHUB_ENV}` file if we don't change the initial value.

Comment (gyohuangxin) [reply .github/workflows/triton-test.yaml:131]: Great, understand.

**Comment by brunomazzottiamd** [comment]
Rebased on top of `main` branch.
